### PR TITLE
Rename plugin.opensslv3, support for OpenSSL 3.0.8

### PR DESCRIPTION
**Pull request by clang-clang-clang**
This is related to https://github.com/coronalabs/corona/issues/534.

## TL;DR; 

This PR fixed the OpenSSL compilation option, re-adapts project using the 3673 template project, and update lua-openssl to the latest version, removing some of the old architecture binaries and providing some of the new architecture binaries. **Since the API of lua-openssl has changed, the name has been changed to `plugin.opensslv3` to avoid breaking compatibility.**

Here are some details,

## Decision

1. Solar2D interface version: Since `lua-openssl` has not been updated for ten years and its interface has changed, the plugin uses a new name to avoid compatibility issues. Therefore, version 3673 from one year ago was selected instead of an earlier version (the interface definition has not substantially changed since 2020).
2. The OpenSSL version: Following [OpenSSL version policy](https://www.openssl.org/policies/releasestrat.html), version 3.0.8 (LTS version, maintained until September 2026) was adapted first since version 1.1.1 will also stop maintenance at the end of 2023. If there are major obstacles, consider upgrading step by step; The adaptation seems to be complete now.
3. Fixed compilation options: The original repository relies on the `sdk-openssl` submodule to compile OpenSSL static libraries. Now all platforms compile on their own (trigger it manually), referencing [x2on/OpenSSL-for-iPhone](https://github.com/x2on/OpenSSL-for-iPhone) and using OpenSSL's Configure script to compile library.
4. The primary task was to understand the plugin source code. The folder structure should remain compatible with the original plugin, so the static library is provided in `xcframework` format. However, the final output retains the old folder structure (.a files, etc.).

## Source code changes

1. Use the same source code dependencies as much as possible. Because `luasocket` is also used in Solar2D core, there may be symbol redefinition issues (such as iOS) for platforms that statically link Solar2D core but not for dynamically used platforms (such as Android). For example, `lua-openssl` carries `auxiliar.c` and `subsidiar.c` from luasocket, but only uses the latter and directly references `auxiliar.c` in luasocket.
2. When compiling on each platform, convert the two lua files in `input_for_lua2c` (`options.lua` is used for the compiler to generate interfaces, not for runtime, so it is deleted) instead of relying on the repository source code.
3. Since the new version of lua-openssl has separated custom modifications from `auxiliar.c/.h`, `corona_auxiliar.c/.h` are no longer needed. The new version of `lua-openssl` has fixed or changed the implementation, so the patch fix in Solar2D for the old version is no longer required.
4. `lua-openssl` uses `lua 5.3` interface to implement it and uses `lua-compat` 5.3 to be compatible with 5.2 5.1. The aforementioned `auxiliar.c` uniformly uses luasocket's, so `auxiliar_getclassudata` in `luasocket` needs to be modified to use `testudata` to make it compatible with both (In fact, `luasocket` is a bit outdated, but it has not been upgraded due to the Solar2D core not being upgraded).
5. The Solar2D project for usage also adapts both plugin version and tested.

### Apple

Use Xcode 14.3 to compile only the 64-bit version.

#### iOS

1. iOS includes iOS (_arm64_), iOS Simulator (_arm64_, _x86_64_), and does not include _Catalyst_. Among them, iOS Simulator _arm64_ is the new binary added this time, but you need to compile the supported `libplayer.a` yourself to use it.
2. Merging the libraries of "different platforms" into a _fat_ library using `lipo` is not recommended, which is one of the reason why `libplayer.a` currently does not support iOS Simulator _arm64_. Instead, _xcframework_ should be used to merge the binaries of "different platforms," and `lipo` should be used to merge the binaries of "different architectures of the same platform."
3. The deployment target was changed to `11.0`, which is reflected in the OpenSSL config to avoid the implicit linking problem caused by not finding libarclite. See [CMake Issues#19990](https://gitlab.kitware.com/cmake/cmake/-/issues/19990).
4. `function.c` prepared for iOS 8 Simulator was deleted.

#### tvOS

1. tvOS includes tvOS (_arm64_), tvOS Simulator (_arm64_, _x86_64_). Among them, tvOS Simulator _arm64_ and _x86_64_ are new binaries added this time (in [plugin.openssl-2016.2883-v2](https://github.com/coronalabs/com.coronalabs-plugin.openssl/releases/tag/v2), tvOS and Simulator have the same binaries), but you need to compile the supported libplayer.a yourself to use it.
2. The deployment target was changed to `9.0`.

#### macOS

1. macOS includes _arm64_, _x86_64_.
2. The deployment target was changed to `10.15`.

### Android

1. Android includes (_armeabi-v7a_, _arm64-v8a_, _x86_, _x86_64_).
2. The NDK uses `r18b`, which I think is a stable version of the NDK before the BIG change (compared to 20 or 22), and we should at least upgrade to this version before Android API 33 arrives.
3. The build project was rewritten using CMake.

### Windows

1. Only win32 is compiled.
2. The 2015 v140xp toolset was used, and you will need [the Visual C++ Redistributable for Visual Studio 2015](https://www.microsoft.com/en-us/download/details.aspx?id=48145) at runtime.

## Repo changes

1. Need to fork the changed `lua-openssl` from the [dianchu/lua-openssl](https://github.com/dianchu/lua-openssl) to the [coronalabs](https://github.com/coronalabs) group.
2. PR for submodule-luasocket was created https://github.com/coronalabs/submodule-luasocket/pull/1.

## CVE status

1. Using `cve-bin-tool` scanning, we have detected 7 vulnerabilities, with 4 being new and the other 3 being unrelated mutt-related vulnerabilities. This marks a significant improvement from the previous detection of 38 vulnerabilities in total.

Thank you.

### CODE REVIEW

**Comment by clang-clang-clang** [comment]
Resolved https://github.com/coronalabs/plugins-source-openssl/issues/1.

**Comment by clang-clang-clang** [comment]
Hi, @Shchvova, will CLA be a problem that prevents PR?

**Comment by Shchvova** [comment]
No, CLA wouldn't be a problem. Hovever, we do not really rename plugins. We make new ones. I will create a copy of a plugin for this purpose
